### ISSUE TITLE
Use scopes & trigger "best guess" docblock from anywhere

### ISSUE THREAD

**Issue by gerardroche** [spadgos/sublime-jsdocs#377]
Currently DocBlockr parses the content after the cursor and uses regular expressions to extract the appropriate content for the docblock. 

For example, most parsers use `\\((?P<args>.*)\\)` to extract a function arguments which means edge cases fail.

``` php
class x
{
    /**|
    public function functionName($a, $b) { return new ClassName(); }

}
```

`\\((?P<args>.*)\\)` matches something like `$a, $b) { return new ClassName(` which is obviously wrong, but the only way to parse these edge cases correctly is with recursive regular expression.

One possible solution is to use syntax scopes, which _are_ recursive, to extract function names and arguments. 

All syntaxes should be similar enough to use this method of extraction, and those that are not should be fixed.

``` php
public function functionName($a, $b) { return new ClassName(); }
                ^^^^^^^^^^^^ source.php meta.function.php  entity.name.function.php
                             ^^^^^^ source.php meta.function.php  meta.function.arguments.php
```

All syntaxes should be able to match the function name with `meta.function entity.name.function` and the arguments with `meta.function meta.function.arguments`.

What could be really good, and I have a very ugly custom macro that does this, is to trigger a docblock from anywhere inside, before or anywhere near the function declaration.

``` php
1  class x
2  {
3      public function a()
4      {
5
6      }
7
8
9
10
11
12    public function b()
13    {
14
15
16     
17 
18
19    }
20 }
```

 When the cursor is on any line 7-19:
- the whitespace on lines 7-11 is removed (could be optional)
- and a docblock to triggered for function b on line 12

If lines 14-18 contained actual code then the docblock that gets triggered might be something else e.g. a variable. However, a specific keybinding for a function docblock would always do the above.  

When on line 6-3:
- a dockblock is triggered for function a

The current context is _key_ to what docblock is triggered.

The above is possible because docblockr moves either forward or backwards based on the current context looking for a syntax scope function declaration or variable or class, etc. and does some best guess work to figure out what the intention might be. Most of the time it's unambiguous, though I'm sure there will be edge cases in which case do nothing or prompt with the options.

It's a lot of work and requires some significant changes.

Feedback, suggestions, questions?

**Comment by 20Tauri** [comment]
[DoxyDoxygen](http://20tauri.free.fr/DoxyDoxygen), an alternative to this great plugin, use non regexp parser. Perhaps you may be interested in it.

Here a possible result of your example using future 0.8.4 version (please note that php support is poor compared to others languages)
```php
class x
{
    /*----------------------------------------------------------------------*//**
     * { description }
     *
     * @param      <type>  $a
     * @param      <type>  $b
     *
     * @return     <type>
     */
    public function functionName($a, $b) { return new ClassName(); }
}
```

In my approach, I haven't choose to use directly tmLanguage:
  - hard to test
  - hard to extend (tmLanguage have to cover all languages aspect).
  - not exactly similar. For example, in Python, inside a function body "meta.function" is unset
  - hard to parse body (even if I not parse PHP body for now ;) ).
  - ...

Your second suggestion is really interesting... I hate _go up_ to insert "/**"...
I just implement it in [DoxyDoxygen](http://20tauri.free.fr/DoxyDoxygen) and it will be available in the next release of this package (0.8.4+).

You don't speak about key mapping. In my package, I will probably map this function on <kbd>Alt</kbd>+<kbd>Q</kbd> (my current key to reparse / wrap comment).

Thanks again for this suggestion and if you test [DoxyDoxygen](http://20tauri.free.fr/DoxyDoxygen), feedbacks are welcomed.

**Comment by 20Tauri** [comment]
A feedback, if you decide to implement in DocBlockR:
   - the main problem is to detect the insertion point
   - tmLanguage have really strange behaviour (example in c++: fn() is a "meta.function", fn/**/() not)
   - as expected, new way of writing comments is really pleasant
